
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="ctDNA/UMI MRD pipeline with formal detection limit analytics and contamination robustness">
      
      
      
        <link rel="canonical" href="https://altalanta.github.io/precise-mrd-mini/tutorials/contamination_analysis/">
      
      
        <link rel="prev" href="../formal_detection_limits/">
      
      
        <link rel="next" href="../../evaluation/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.20">
    
    
      
        <title>Contamination Analysis - Precise MRD - Detection Limit Analytics</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.e53b48f4.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#contamination-analysis-tutorial" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Precise MRD - Detection Limit Analytics" class="md-header__button md-logo" aria-label="Precise MRD - Detection Limit Analytics" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Precise MRD - Detection Limit Analytics
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Contamination Analysis
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="blue" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="blue" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/altalanta/precise-mrd-mini" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    altalanta/precise-mrd-mini
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../getting-started/" class="md-tabs__link">
        
  
  
    
  
  Getting Started

      </a>
    </li>
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../formal_detection_limits/" class="md-tabs__link">
          
  
  
  Tutorials

        </a>
      </li>
    
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../evaluation/" class="md-tabs__link">
        
  
  
    
  
  Evaluation

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../contamination/" class="md-tabs__link">
        
  
  
    
  
  Contamination Analysis

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../statistical-validation/" class="md-tabs__link">
        
  
  
    
  
  Statistical Validation

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../reproducibility/" class="md-tabs__link">
        
  
  
    
  
  Reproducibility

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../api-reference/" class="md-tabs__link">
        
  
  
    
  
  API Reference

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Precise MRD - Detection Limit Analytics" class="md-nav__button md-logo" aria-label="Precise MRD - Detection Limit Analytics" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Precise MRD - Detection Limit Analytics
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/altalanta/precise-mrd-mini" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    altalanta/precise-mrd-mini
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../getting-started/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Getting Started
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Tutorials
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../formal_detection_limits/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Formal Detection Limits
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Contamination Analysis
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Contamination Analysis
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#why-contamination-matters" class="md-nav__link">
    <span class="md-ellipsis">
      Why Contamination Matters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    <span class="md-ellipsis">
      Learning Objectives
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#1-types-of-contamination" class="md-nav__link">
    <span class="md-ellipsis">
      1. Types of Contamination
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Types of Contamination">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#index-hopping" class="md-nav__link">
    <span class="md-ellipsis">
      Index Hopping
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sample-carryover" class="md-nav__link">
    <span class="md-ellipsis">
      Sample Carryover
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#environmental-contamination" class="md-nav__link">
    <span class="md-ellipsis">
      Environmental Contamination
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-simulating-contamination-scenarios" class="md-nav__link">
    <span class="md-ellipsis">
      2. Simulating Contamination Scenarios
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-impact-on-detection-limits" class="md-nav__link">
    <span class="md-ellipsis">
      3. Impact on Detection Limits
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-mitigation-strategies" class="md-nav__link">
    <span class="md-ellipsis">
      4. Mitigation Strategies
    </span>
  </a>
  
    <nav class="md-nav" aria-label="4. Mitigation Strategies">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#unique-molecular-identifiers-umis" class="md-nav__link">
    <span class="md-ellipsis">
      Unique Molecular Identifiers (UMIs)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5-using-precise-mrd-contamination-commands" class="md-nav__link">
    <span class="md-ellipsis">
      5. Using Precise MRD Contamination Commands
    </span>
  </a>
  
    <nav class="md-nav" aria-label="5. Using Precise MRD Contamination Commands">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#basic-contamination-evaluation" class="md-nav__link">
    <span class="md-ellipsis">
      Basic Contamination Evaluation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#index-hopping-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Index Hopping Analysis
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mitigation-strategy-testing" class="md-nav__link">
    <span class="md-ellipsis">
      Mitigation Strategy Testing
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#6-interpreting-contamination-results" class="md-nav__link">
    <span class="md-ellipsis">
      6. Interpreting Contamination Results
    </span>
  </a>
  
    <nav class="md-nav" aria-label="6. Interpreting Contamination Results">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#key-metrics-to-monitor" class="md-nav__link">
    <span class="md-ellipsis">
      Key Metrics to Monitor
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#acceptable-contamination-levels" class="md-nav__link">
    <span class="md-ellipsis">
      Acceptable Contamination Levels
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quality-control-thresholds" class="md-nav__link">
    <span class="md-ellipsis">
      Quality Control Thresholds
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#7-best-practices-for-contamination-control" class="md-nav__link">
    <span class="md-ellipsis">
      7. Best Practices for Contamination Control
    </span>
  </a>
  
    <nav class="md-nav" aria-label="7. Best Practices for Contamination Control">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#laboratory-practices" class="md-nav__link">
    <span class="md-ellipsis">
      Laboratory Practices
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#computational-mitigation" class="md-nav__link">
    <span class="md-ellipsis">
      Computational Mitigation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#conclusion" class="md-nav__link">
    <span class="md-ellipsis">
      Conclusion
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Conclusion">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#next-steps" class="md-nav__link">
    <span class="md-ellipsis">
      Next Steps
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../evaluation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Evaluation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../contamination/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Contamination Analysis
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../statistical-validation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Statistical Validation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../reproducibility/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Reproducibility
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api-reference/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    API Reference
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#why-contamination-matters" class="md-nav__link">
    <span class="md-ellipsis">
      Why Contamination Matters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    <span class="md-ellipsis">
      Learning Objectives
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#1-types-of-contamination" class="md-nav__link">
    <span class="md-ellipsis">
      1. Types of Contamination
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Types of Contamination">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#index-hopping" class="md-nav__link">
    <span class="md-ellipsis">
      Index Hopping
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sample-carryover" class="md-nav__link">
    <span class="md-ellipsis">
      Sample Carryover
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#environmental-contamination" class="md-nav__link">
    <span class="md-ellipsis">
      Environmental Contamination
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-simulating-contamination-scenarios" class="md-nav__link">
    <span class="md-ellipsis">
      2. Simulating Contamination Scenarios
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-impact-on-detection-limits" class="md-nav__link">
    <span class="md-ellipsis">
      3. Impact on Detection Limits
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-mitigation-strategies" class="md-nav__link">
    <span class="md-ellipsis">
      4. Mitigation Strategies
    </span>
  </a>
  
    <nav class="md-nav" aria-label="4. Mitigation Strategies">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#unique-molecular-identifiers-umis" class="md-nav__link">
    <span class="md-ellipsis">
      Unique Molecular Identifiers (UMIs)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5-using-precise-mrd-contamination-commands" class="md-nav__link">
    <span class="md-ellipsis">
      5. Using Precise MRD Contamination Commands
    </span>
  </a>
  
    <nav class="md-nav" aria-label="5. Using Precise MRD Contamination Commands">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#basic-contamination-evaluation" class="md-nav__link">
    <span class="md-ellipsis">
      Basic Contamination Evaluation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#index-hopping-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Index Hopping Analysis
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mitigation-strategy-testing" class="md-nav__link">
    <span class="md-ellipsis">
      Mitigation Strategy Testing
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#6-interpreting-contamination-results" class="md-nav__link">
    <span class="md-ellipsis">
      6. Interpreting Contamination Results
    </span>
  </a>
  
    <nav class="md-nav" aria-label="6. Interpreting Contamination Results">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#key-metrics-to-monitor" class="md-nav__link">
    <span class="md-ellipsis">
      Key Metrics to Monitor
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#acceptable-contamination-levels" class="md-nav__link">
    <span class="md-ellipsis">
      Acceptable Contamination Levels
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quality-control-thresholds" class="md-nav__link">
    <span class="md-ellipsis">
      Quality Control Thresholds
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#7-best-practices-for-contamination-control" class="md-nav__link">
    <span class="md-ellipsis">
      7. Best Practices for Contamination Control
    </span>
  </a>
  
    <nav class="md-nav" aria-label="7. Best Practices for Contamination Control">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#laboratory-practices" class="md-nav__link">
    <span class="md-ellipsis">
      Laboratory Practices
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#computational-mitigation" class="md-nav__link">
    <span class="md-ellipsis">
      Computational Mitigation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#conclusion" class="md-nav__link">
    <span class="md-ellipsis">
      Conclusion
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Conclusion">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#next-steps" class="md-nav__link">
    <span class="md-ellipsis">
      Next Steps
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="contamination-analysis-tutorial">Contamination Analysis Tutorial</h1>
<p>This tutorial demonstrates how to analyze and mitigate contamination in ctDNA/MRD analysis using Precise MRD. Contamination is a critical concern in liquid biopsy assays where even small amounts of cross-sample contamination can lead to false positive results.</p>
<h2 id="why-contamination-matters">Why Contamination Matters</h2>
<p>In ctDNA analysis, contamination can occur through:</p>
<ul>
<li><strong>Index hopping</strong>: Misassigned reads between multiplexed samples</li>
<li><strong>Sample carryover</strong>: Residual DNA from previous samples</li>
<li><strong>PCR artifacts</strong>: Template switching or jumping</li>
<li><strong>Environmental contamination</strong>: External DNA sources</li>
</ul>
<p>Even 0.1% contamination can be clinically significant when dealing with MRD detection at 0.01% allele frequencies.</p>
<h2 id="learning-objectives">Learning Objectives</h2>
<p>By the end of this tutorial, you will be able to:
- Understand different types of contamination in ctDNA analysis
- Simulate contamination scenarios using Precise MRD
- Calculate contamination impact on detection limits
- Implement contamination mitigation strategies
- Interpret contamination analysis results</p>
<h2 id="1-types-of-contamination">1. Types of Contamination</h2>
<h3 id="index-hopping">Index Hopping</h3>
<p>Index hopping occurs during sequencing when reads are misassigned between samples in multiplexed libraries. This is particularly problematic for MRD analysis where we're looking for rare variants.</p>
<h3 id="sample-carryover">Sample Carryover</h3>
<p>Sample carryover happens during sample preparation when residual DNA from one sample contaminates subsequent samples. This is especially concerning in automated liquid handling systems.</p>
<h3 id="environmental-contamination">Environmental Contamination</h3>
<p>Environmental contamination occurs when external DNA sources (lab personnel, equipment, reagents) introduce artifacts into the analysis.</p>
<h2 id="2-simulating-contamination-scenarios">2. Simulating Contamination Scenarios</h2>
<p>Let's start by simulating different contamination scenarios and understanding their impact:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="c1"># Set reproducible seed</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="c1"># Define contamination scenarios</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="n">contamination_rates</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.0001</span><span class="p">,</span> <span class="mf">0.001</span><span class="p">,</span> <span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">]</span>  <span class="c1"># 0.01% to 10%</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="n">true_allele_frequencies</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.001</span><span class="p">,</span> <span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">]</span>  <span class="c1"># 0% to 10%</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="n">sequencing_depth</span> <span class="o">=</span> <span class="mi">10000</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Contamination Analysis Parameters:&quot;</span><span class="p">)</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Contamination rates: </span><span class="si">{</span><span class="n">contamination_rates</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;True allele frequencies: </span><span class="si">{</span><span class="n">true_allele_frequencies</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Sequencing depth: </span><span class="si">{</span><span class="n">sequencing_depth</span><span class="si">}</span><span class="s2">x&quot;</span><span class="p">)</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="nb">print</span><span class="p">()</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="c1"># Simulate contamination effects</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="n">contamination_results</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="k">for</span> <span class="n">true_af</span> <span class="ow">in</span> <span class="n">true_allele_frequencies</span><span class="p">:</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="k">for</span> <span class="n">contam_rate</span> <span class="ow">in</span> <span class="n">contamination_rates</span><span class="p">:</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>        <span class="c1"># Simulate observed allele frequency with contamination</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>        <span class="c1"># True signal + contamination noise</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="n">observed_af</span> <span class="o">=</span> <span class="n">true_af</span> <span class="o">+</span> <span class="n">contam_rate</span> <span class="o">*</span> <span class="mf">0.001</span>  <span class="c1"># Assume contaminant has 0.1% AF</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>        <span class="c1"># Add sampling noise (binomial distribution)</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>        <span class="n">observed_mutants</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">binomial</span><span class="p">(</span><span class="n">sequencing_depth</span><span class="p">,</span> <span class="n">observed_af</span><span class="p">)</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>        <span class="n">observed_af_sampled</span> <span class="o">=</span> <span class="n">observed_mutants</span> <span class="o">/</span> <span class="n">sequencing_depth</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>        <span class="n">contamination_results</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>            <span class="s1">&#39;true_af&#39;</span><span class="p">:</span> <span class="n">true_af</span><span class="p">,</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>            <span class="s1">&#39;contamination_rate&#39;</span><span class="p">:</span> <span class="n">contam_rate</span><span class="p">,</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>            <span class="s1">&#39;expected_af&#39;</span><span class="p">:</span> <span class="n">observed_af</span><span class="p">,</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>            <span class="s1">&#39;observed_af&#39;</span><span class="p">:</span> <span class="n">observed_af_sampled</span><span class="p">,</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>            <span class="s1">&#39;false_positive_risk&#39;</span><span class="p">:</span> <span class="n">observed_af_sampled</span> <span class="o">&gt;</span> <span class="mf">0.001</span> <span class="k">if</span> <span class="n">true_af</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="kc">False</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>        <span class="p">})</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="n">df_contamination</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">contamination_results</span><span class="p">)</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sample Results:&quot;</span><span class="p">)</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="nb">print</span><span class="p">(</span><span class="n">df_contamination</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span><span class="o">.</span><span class="n">to_string</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
</code></pre></div>
<h2 id="3-impact-on-detection-limits">3. Impact on Detection Limits</h2>
<p>Contamination directly affects our ability to detect true MRD signals. Let's analyze how different contamination rates impact LoD:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="c1"># Analyze false positive rates</span>
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a><span class="n">false_positive_analysis</span> <span class="o">=</span> <span class="n">df_contamination</span><span class="p">[</span><span class="n">df_contamination</span><span class="p">[</span><span class="s1">&#39;true_af&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mf">0.0</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a><span class="n">false_positive_summary</span> <span class="o">=</span> <span class="n">false_positive_analysis</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;contamination_rate&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span>
<a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a>    <span class="s1">&#39;false_positive_risk&#39;</span><span class="p">:</span> <span class="s1">&#39;mean&#39;</span>
<a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a><span class="p">})</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a>
<a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;False Positive Risk by Contamination Rate:&quot;</span><span class="p">)</span>
<a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a><span class="nb">print</span><span class="p">(</span><span class="n">false_positive_summary</span><span class="o">.</span><span class="n">to_string</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">float_format</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%.4f</span><span class="s1">&#39;</span><span class="p">))</span>
<a id="__codelineno-1-9" name="__codelineno-1-9" href="#__codelineno-1-9"></a>
<a id="__codelineno-1-10" name="__codelineno-1-10" href="#__codelineno-1-10"></a><span class="c1"># Visualize contamination impact</span>
<a id="__codelineno-1-11" name="__codelineno-1-11" href="#__codelineno-1-11"></a><span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<a id="__codelineno-1-12" name="__codelineno-1-12" href="#__codelineno-1-12"></a>
<a id="__codelineno-1-13" name="__codelineno-1-13" href="#__codelineno-1-13"></a><span class="c1"># False positive rates</span>
<a id="__codelineno-1-14" name="__codelineno-1-14" href="#__codelineno-1-14"></a><span class="n">ax1</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">false_positive_summary</span><span class="p">[</span><span class="s1">&#39;contamination_rate&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span>
<a id="__codelineno-1-15" name="__codelineno-1-15" href="#__codelineno-1-15"></a>        <span class="n">false_positive_summary</span><span class="p">[</span><span class="s1">&#39;false_positive_risk&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span>
<a id="__codelineno-1-16" name="__codelineno-1-16" href="#__codelineno-1-16"></a>        <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
<a id="__codelineno-1-17" name="__codelineno-1-17" href="#__codelineno-1-17"></a><span class="n">ax1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Contamination Rate (%)&#39;</span><span class="p">)</span>
<a id="__codelineno-1-18" name="__codelineno-1-18" href="#__codelineno-1-18"></a><span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;False Positive Risk (%)&#39;</span><span class="p">)</span>
<a id="__codelineno-1-19" name="__codelineno-1-19" href="#__codelineno-1-19"></a><span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;False Positive Risk vs Contamination Rate&#39;</span><span class="p">)</span>
<a id="__codelineno-1-20" name="__codelineno-1-20" href="#__codelineno-1-20"></a><span class="n">ax1</span><span class="o">.</span><span class="n">set_yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
<a id="__codelineno-1-21" name="__codelineno-1-21" href="#__codelineno-1-21"></a><span class="n">ax1</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
<a id="__codelineno-1-22" name="__codelineno-1-22" href="#__codelineno-1-22"></a>
<a id="__codelineno-1-23" name="__codelineno-1-23" href="#__codelineno-1-23"></a><span class="c1"># Observed vs expected AF</span>
<a id="__codelineno-1-24" name="__codelineno-1-24" href="#__codelineno-1-24"></a><span class="k">for</span> <span class="n">true_af</span> <span class="ow">in</span> <span class="n">true_allele_frequencies</span><span class="p">:</span>
<a id="__codelineno-1-25" name="__codelineno-1-25" href="#__codelineno-1-25"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">df_contamination</span><span class="p">[</span><span class="n">df_contamination</span><span class="p">[</span><span class="s1">&#39;true_af&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">true_af</span><span class="p">]</span>
<a id="__codelineno-1-26" name="__codelineno-1-26" href="#__codelineno-1-26"></a>    <span class="n">ax2</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;contamination_rate&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;observed_af&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span>
<a id="__codelineno-1-27" name="__codelineno-1-27" href="#__codelineno-1-27"></a>               <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;True AF: </span><span class="si">{</span><span class="n">true_af</span><span class="o">*</span><span class="mi">100</span><span class="si">:</span><span class="s1">.1f</span><span class="si">}</span><span class="s1">%&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
<a id="__codelineno-1-28" name="__codelineno-1-28" href="#__codelineno-1-28"></a>
<a id="__codelineno-1-29" name="__codelineno-1-29" href="#__codelineno-1-29"></a><span class="n">ax2</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Contamination Rate (%)&#39;</span><span class="p">)</span>
<a id="__codelineno-1-30" name="__codelineno-1-30" href="#__codelineno-1-30"></a><span class="n">ax2</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Observed Allele Frequency (%)&#39;</span><span class="p">)</span>
<a id="__codelineno-1-31" name="__codelineno-1-31" href="#__codelineno-1-31"></a><span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Observed AF vs Contamination Rate&#39;</span><span class="p">)</span>
<a id="__codelineno-1-32" name="__codelineno-1-32" href="#__codelineno-1-32"></a><span class="n">ax2</span><span class="o">.</span><span class="n">set_yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
<a id="__codelineno-1-33" name="__codelineno-1-33" href="#__codelineno-1-33"></a><span class="n">ax2</span><span class="o">.</span><span class="n">set_xscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
<a id="__codelineno-1-34" name="__codelineno-1-34" href="#__codelineno-1-34"></a><span class="n">ax2</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<a id="__codelineno-1-35" name="__codelineno-1-35" href="#__codelineno-1-35"></a><span class="n">ax2</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
<a id="__codelineno-1-36" name="__codelineno-1-36" href="#__codelineno-1-36"></a>
<a id="__codelineno-1-37" name="__codelineno-1-37" href="#__codelineno-1-37"></a><span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<a id="__codelineno-1-38" name="__codelineno-1-38" href="#__codelineno-1-38"></a><span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div>
<h2 id="4-mitigation-strategies">4. Mitigation Strategies</h2>
<h3 id="unique-molecular-identifiers-umis">Unique Molecular Identifiers (UMIs)</h3>
<p>UMIs help distinguish true mutations from contamination by providing molecular barcodes:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="c1"># Simulate UMI-based contamination detection</span>
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a><span class="k">def</span><span class="w"> </span><span class="nf">simulate_umi_contamination_detection</span><span class="p">(</span><span class="n">n_umis_per_molecule</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">contamination_rate</span><span class="o">=</span><span class="mf">0.001</span><span class="p">):</span>
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;Simulate how UMIs help detect contamination&quot;&quot;&quot;</span>
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a>
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a>    <span class="c1"># True sample: 0.01% AF with 5 UMI families</span>
<a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a>    <span class="n">true_molecules</span> <span class="o">=</span> <span class="mi">100000</span>  <span class="c1"># Total molecules sequenced</span>
<a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a>    <span class="n">true_mutant_molecules</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">true_molecules</span> <span class="o">*</span> <span class="mf">0.0001</span><span class="p">)</span>  <span class="c1"># 0.01% AF</span>
<a id="__codelineno-2-8" name="__codelineno-2-8" href="#__codelineno-2-8"></a>
<a id="__codelineno-2-9" name="__codelineno-2-9" href="#__codelineno-2-9"></a>    <span class="c1"># Generate UMI families for true mutants</span>
<a id="__codelineno-2-10" name="__codelineno-2-10" href="#__codelineno-2-10"></a>    <span class="n">true_umi_families</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">true_mutant_molecules</span><span class="p">)</span>  <span class="c1"># Average 3 reads per family</span>
<a id="__codelineno-2-11" name="__codelineno-2-11" href="#__codelineno-2-11"></a>
<a id="__codelineno-2-12" name="__codelineno-2-12" href="#__codelineno-2-12"></a>    <span class="c1"># Contamination: 0.1% AF with different UMI pattern</span>
<a id="__codelineno-2-13" name="__codelineno-2-13" href="#__codelineno-2-13"></a>    <span class="n">contaminant_molecules</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">true_molecules</span> <span class="o">*</span> <span class="n">contamination_rate</span><span class="p">)</span>
<a id="__codelineno-2-14" name="__codelineno-2-14" href="#__codelineno-2-14"></a>    <span class="n">contaminant_umi_families</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">contaminant_molecules</span><span class="p">)</span>  <span class="c1"># Average 1 read per family</span>
<a id="__codelineno-2-15" name="__codelineno-2-15" href="#__codelineno-2-15"></a>
<a id="__codelineno-2-16" name="__codelineno-2-16" href="#__codelineno-2-16"></a>    <span class="c1"># Analysis</span>
<a id="__codelineno-2-17" name="__codelineno-2-17" href="#__codelineno-2-17"></a>    <span class="n">total_observed_mutants</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">true_umi_families</span><span class="p">)</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">contaminant_umi_families</span><span class="p">)</span>
<a id="__codelineno-2-18" name="__codelineno-2-18" href="#__codelineno-2-18"></a>    <span class="n">total_umi_families</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">true_umi_families</span><span class="p">)</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">contaminant_umi_families</span><span class="p">)</span>
<a id="__codelineno-2-19" name="__codelineno-2-19" href="#__codelineno-2-19"></a>
<a id="__codelineno-2-20" name="__codelineno-2-20" href="#__codelineno-2-20"></a>    <span class="c1"># UMI-based AF estimate (more accurate)</span>
<a id="__codelineno-2-21" name="__codelineno-2-21" href="#__codelineno-2-21"></a>    <span class="n">umi_based_af</span> <span class="o">=</span> <span class="n">total_umi_families</span> <span class="o">/</span> <span class="n">true_molecules</span>
<a id="__codelineno-2-22" name="__codelineno-2-22" href="#__codelineno-2-22"></a>
<a id="__codelineno-2-23" name="__codelineno-2-23" href="#__codelineno-2-23"></a>    <span class="c1"># Read-based AF estimate (contaminated)</span>
<a id="__codelineno-2-24" name="__codelineno-2-24" href="#__codelineno-2-24"></a>    <span class="n">read_based_af</span> <span class="o">=</span> <span class="n">total_observed_mutants</span> <span class="o">/</span> <span class="p">(</span><span class="n">true_molecules</span> <span class="o">*</span> <span class="n">n_umis_per_molecule</span><span class="p">)</span>
<a id="__codelineno-2-25" name="__codelineno-2-25" href="#__codelineno-2-25"></a>
<a id="__codelineno-2-26" name="__codelineno-2-26" href="#__codelineno-2-26"></a>    <span class="k">return</span> <span class="p">{</span>
<a id="__codelineno-2-27" name="__codelineno-2-27" href="#__codelineno-2-27"></a>        <span class="s1">&#39;true_af&#39;</span><span class="p">:</span> <span class="mf">0.0001</span><span class="p">,</span>
<a id="__codelineno-2-28" name="__codelineno-2-28" href="#__codelineno-2-28"></a>        <span class="s1">&#39;umi_based_af&#39;</span><span class="p">:</span> <span class="n">umi_based_af</span><span class="p">,</span>
<a id="__codelineno-2-29" name="__codelineno-2-29" href="#__codelineno-2-29"></a>        <span class="s1">&#39;read_based_af&#39;</span><span class="p">:</span> <span class="n">read_based_af</span><span class="p">,</span>
<a id="__codelineno-2-30" name="__codelineno-2-30" href="#__codelineno-2-30"></a>        <span class="s1">&#39;contamination_detected&#39;</span><span class="p">:</span> <span class="n">umi_based_af</span> <span class="o">&lt;</span> <span class="n">read_based_af</span> <span class="o">*</span> <span class="mf">0.5</span>
<a id="__codelineno-2-31" name="__codelineno-2-31" href="#__codelineno-2-31"></a>    <span class="p">}</span>
<a id="__codelineno-2-32" name="__codelineno-2-32" href="#__codelineno-2-32"></a>
<a id="__codelineno-2-33" name="__codelineno-2-33" href="#__codelineno-2-33"></a><span class="c1"># Test UMI effectiveness</span>
<a id="__codelineno-2-34" name="__codelineno-2-34" href="#__codelineno-2-34"></a><span class="n">umi_results</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-2-35" name="__codelineno-2-35" href="#__codelineno-2-35"></a><span class="k">for</span> <span class="n">contam_rate</span> <span class="ow">in</span> <span class="p">[</span><span class="mf">0.0001</span><span class="p">,</span> <span class="mf">0.001</span><span class="p">,</span> <span class="mf">0.01</span><span class="p">]:</span>
<a id="__codelineno-2-36" name="__codelineno-2-36" href="#__codelineno-2-36"></a>    <span class="n">result</span> <span class="o">=</span> <span class="n">simulate_umi_contamination_detection</span><span class="p">(</span><span class="n">contamination_rate</span><span class="o">=</span><span class="n">contam_rate</span><span class="p">)</span>
<a id="__codelineno-2-37" name="__codelineno-2-37" href="#__codelineno-2-37"></a>    <span class="n">umi_results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
<a id="__codelineno-2-38" name="__codelineno-2-38" href="#__codelineno-2-38"></a>
<a id="__codelineno-2-39" name="__codelineno-2-39" href="#__codelineno-2-39"></a><span class="n">df_umi</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">umi_results</span><span class="p">)</span>
<a id="__codelineno-2-40" name="__codelineno-2-40" href="#__codelineno-2-40"></a><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;UMI-based Contamination Detection:&quot;</span><span class="p">)</span>
<a id="__codelineno-2-41" name="__codelineno-2-41" href="#__codelineno-2-41"></a><span class="nb">print</span><span class="p">(</span><span class="n">df_umi</span><span class="o">.</span><span class="n">to_string</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">float_format</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%.6f</span><span class="s1">&#39;</span><span class="p">))</span>
</code></pre></div>
<h2 id="5-using-precise-mrd-contamination-commands">5. Using Precise MRD Contamination Commands</h2>
<p>Now let's use the actual Precise MRD commands for comprehensive contamination analysis:</p>
<h3 id="basic-contamination-evaluation">Basic Contamination Evaluation</h3>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="c1"># Evaluate contamination robustness across different scenarios</span>
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>precise-mrd<span class="w"> </span>eval-contamination<span class="w"> </span>--output<span class="w"> </span>reports/contamination_analysis.json
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a>
<a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a><span class="c1"># Generate contamination sensitivity heatmap</span>
<a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a>precise-mrd<span class="w"> </span>eval-contamination<span class="w"> </span>--heatmap<span class="w"> </span>--output<span class="w"> </span>reports/contamination_heatmap.png
<a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a>
<a id="__codelineno-3-7" name="__codelineno-3-7" href="#__codelineno-3-7"></a><span class="c1"># Test specific contamination rates</span>
<a id="__codelineno-3-8" name="__codelineno-3-8" href="#__codelineno-3-8"></a>precise-mrd<span class="w"> </span>eval-contamination<span class="w"> </span>--rates<span class="w"> </span><span class="m">0</span>.001<span class="w"> </span><span class="m">0</span>.01<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span>--output<span class="w"> </span>reports/specific_contamination.json
</code></pre></div>
<h3 id="index-hopping-analysis">Index Hopping Analysis</h3>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="c1"># Simulate and analyze index hopping effects</span>
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a>precise-mrd<span class="w"> </span>eval-contamination<span class="w"> </span>--index-hopping<span class="w"> </span>--hop-rate<span class="w"> </span><span class="m">0</span>.001<span class="w"> </span>--output<span class="w"> </span>reports/index_hopping.json
<a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a>
<a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a><span class="c1"># Multi-sample contamination analysis</span>
<a id="__codelineno-4-5" name="__codelineno-4-5" href="#__codelineno-4-5"></a>precise-mrd<span class="w"> </span>eval-contamination<span class="w"> </span>--multiplex<span class="w"> </span><span class="m">96</span><span class="w"> </span>--output<span class="w"> </span>reports/multiplex_contamination.json
</code></pre></div>
<h3 id="mitigation-strategy-testing">Mitigation Strategy Testing</h3>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a><span class="c1"># Test UMI-based contamination detection</span>
<a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a>precise-mrd<span class="w"> </span>eval-contamination<span class="w"> </span>--umi-families<span class="w"> </span>--min-family-size<span class="w"> </span><span class="m">3</span><span class="w"> </span>--output<span class="w"> </span>reports/umi_contamination.json
<a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a>
<a id="__codelineno-5-4" name="__codelineno-5-4" href="#__codelineno-5-4"></a><span class="c1"># Evaluate decontamination strategies</span>
<a id="__codelineno-5-5" name="__codelineno-5-5" href="#__codelineno-5-5"></a>precise-mrd<span class="w"> </span>eval-contamination<span class="w"> </span>--decontamination<span class="w"> </span>--strategy<span class="w"> </span>consensus<span class="w"> </span>--output<span class="w"> </span>reports/decontaminated.json
</code></pre></div>
<h2 id="6-interpreting-contamination-results">6. Interpreting Contamination Results</h2>
<h3 id="key-metrics-to-monitor">Key Metrics to Monitor</h3>
<ol>
<li><strong>False Positive Rate</strong>: Percentage of negative samples incorrectly called positive</li>
<li><strong>Sensitivity Loss</strong>: Reduction in true positive detection due to contamination</li>
<li><strong>Quantification Accuracy</strong>: How well we measure true allele frequencies</li>
<li><strong>UMI Family Distribution</strong>: Pattern of read distribution across molecular families</li>
</ol>
<h3 id="acceptable-contamination-levels">Acceptable Contamination Levels</h3>
<ul>
<li><strong>Research applications</strong>: &lt; 0.1% contamination acceptable</li>
<li><strong>Clinical MRD</strong>: &lt; 0.01% contamination required</li>
<li><strong>Ultra-sensitive MRD</strong>: &lt; 0.001% contamination target</li>
</ul>
<h3 id="quality-control-thresholds">Quality Control Thresholds</h3>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="c1"># Example QC thresholds</span>
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a><span class="n">QC_THRESHOLDS</span> <span class="o">=</span> <span class="p">{</span>
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a>    <span class="s1">&#39;max_contamination_rate&#39;</span><span class="p">:</span> <span class="mf">0.001</span><span class="p">,</span>  <span class="c1"># 0.1%</span>
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a>    <span class="s1">&#39;min_umi_family_size&#39;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
<a id="__codelineno-6-5" name="__codelineno-6-5" href="#__codelineno-6-5"></a>    <span class="s1">&#39;max_index_hopping_rate&#39;</span><span class="p">:</span> <span class="mf">0.0005</span><span class="p">,</span>  <span class="c1"># 0.05%</span>
<a id="__codelineno-6-6" name="__codelineno-6-6" href="#__codelineno-6-6"></a>    <span class="s1">&#39;min_detection_efficiency&#39;</span><span class="p">:</span> <span class="mf">0.95</span>   <span class="c1"># 95%</span>
<a id="__codelineno-6-7" name="__codelineno-6-7" href="#__codelineno-6-7"></a><span class="p">}</span>
<a id="__codelineno-6-8" name="__codelineno-6-8" href="#__codelineno-6-8"></a>
<a id="__codelineno-6-9" name="__codelineno-6-9" href="#__codelineno-6-9"></a><span class="k">def</span><span class="w"> </span><span class="nf">validate_sample_quality</span><span class="p">(</span><span class="n">contamination_results</span><span class="p">,</span> <span class="n">qc_thresholds</span><span class="p">):</span>
<a id="__codelineno-6-10" name="__codelineno-6-10" href="#__codelineno-6-10"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;Validate if sample meets quality criteria&quot;&quot;&quot;</span>
<a id="__codelineno-6-11" name="__codelineno-6-11" href="#__codelineno-6-11"></a>
<a id="__codelineno-6-12" name="__codelineno-6-12" href="#__codelineno-6-12"></a>    <span class="n">validation_results</span> <span class="o">=</span> <span class="p">{}</span>
<a id="__codelineno-6-13" name="__codelineno-6-13" href="#__codelineno-6-13"></a>
<a id="__codelineno-6-14" name="__codelineno-6-14" href="#__codelineno-6-14"></a>    <span class="c1"># Check contamination rate</span>
<a id="__codelineno-6-15" name="__codelineno-6-15" href="#__codelineno-6-15"></a>    <span class="n">validation_results</span><span class="p">[</span><span class="s1">&#39;contamination_ok&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">contamination_results</span><span class="p">[</span><span class="s1">&#39;rate&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">qc_thresholds</span><span class="p">[</span><span class="s1">&#39;max_contamination_rate&#39;</span><span class="p">]</span>
<a id="__codelineno-6-16" name="__codelineno-6-16" href="#__codelineno-6-16"></a>
<a id="__codelineno-6-17" name="__codelineno-6-17" href="#__codelineno-6-17"></a>    <span class="c1"># Check UMI quality</span>
<a id="__codelineno-6-18" name="__codelineno-6-18" href="#__codelineno-6-18"></a>    <span class="n">validation_results</span><span class="p">[</span><span class="s1">&#39;umi_ok&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">contamination_results</span><span class="p">[</span><span class="s1">&#39;mean_family_size&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">qc_thresholds</span><span class="p">[</span><span class="s1">&#39;min_umi_family_size&#39;</span><span class="p">]</span>
<a id="__codelineno-6-19" name="__codelineno-6-19" href="#__codelineno-6-19"></a>
<a id="__codelineno-6-20" name="__codelineno-6-20" href="#__codelineno-6-20"></a>    <span class="c1"># Check detection efficiency</span>
<a id="__codelineno-6-21" name="__codelineno-6-21" href="#__codelineno-6-21"></a>    <span class="n">validation_results</span><span class="p">[</span><span class="s1">&#39;efficiency_ok&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">contamination_results</span><span class="p">[</span><span class="s1">&#39;detection_efficiency&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">qc_thresholds</span><span class="p">[</span><span class="s1">&#39;min_detection_efficiency&#39;</span><span class="p">]</span>
<a id="__codelineno-6-22" name="__codelineno-6-22" href="#__codelineno-6-22"></a>
<a id="__codelineno-6-23" name="__codelineno-6-23" href="#__codelineno-6-23"></a>    <span class="c1"># Overall quality score</span>
<a id="__codelineno-6-24" name="__codelineno-6-24" href="#__codelineno-6-24"></a>    <span class="n">validation_results</span><span class="p">[</span><span class="s1">&#39;quality_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">validation_results</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">validation_results</span><span class="p">)</span>
<a id="__codelineno-6-25" name="__codelineno-6-25" href="#__codelineno-6-25"></a>
<a id="__codelineno-6-26" name="__codelineno-6-26" href="#__codelineno-6-26"></a>    <span class="k">return</span> <span class="n">validation_results</span>
<a id="__codelineno-6-27" name="__codelineno-6-27" href="#__codelineno-6-27"></a>
<a id="__codelineno-6-28" name="__codelineno-6-28" href="#__codelineno-6-28"></a><span class="c1"># Example validation</span>
<a id="__codelineno-6-29" name="__codelineno-6-29" href="#__codelineno-6-29"></a><span class="n">sample_results</span> <span class="o">=</span> <span class="p">{</span>
<a id="__codelineno-6-30" name="__codelineno-6-30" href="#__codelineno-6-30"></a>    <span class="s1">&#39;rate&#39;</span><span class="p">:</span> <span class="mf">0.0005</span><span class="p">,</span>
<a id="__codelineno-6-31" name="__codelineno-6-31" href="#__codelineno-6-31"></a>    <span class="s1">&#39;mean_family_size&#39;</span><span class="p">:</span> <span class="mf">4.2</span><span class="p">,</span>
<a id="__codelineno-6-32" name="__codelineno-6-32" href="#__codelineno-6-32"></a>    <span class="s1">&#39;detection_efficiency&#39;</span><span class="p">:</span> <span class="mf">0.97</span>
<a id="__codelineno-6-33" name="__codelineno-6-33" href="#__codelineno-6-33"></a><span class="p">}</span>
<a id="__codelineno-6-34" name="__codelineno-6-34" href="#__codelineno-6-34"></a>
<a id="__codelineno-6-35" name="__codelineno-6-35" href="#__codelineno-6-35"></a><span class="n">validation</span> <span class="o">=</span> <span class="n">validate_sample_quality</span><span class="p">(</span><span class="n">sample_results</span><span class="p">,</span> <span class="n">QC_THRESHOLDS</span><span class="p">)</span>
<a id="__codelineno-6-36" name="__codelineno-6-36" href="#__codelineno-6-36"></a><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sample Quality Validation:&quot;</span><span class="p">)</span>
<a id="__codelineno-6-37" name="__codelineno-6-37" href="#__codelineno-6-37"></a><span class="k">for</span> <span class="n">metric</span><span class="p">,</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">validation</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-6-38" name="__codelineno-6-38" href="#__codelineno-6-38"></a>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">result</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div>
<h2 id="7-best-practices-for-contamination-control">7. Best Practices for Contamination Control</h2>
<h3 id="laboratory-practices">Laboratory Practices</h3>
<ol>
<li><strong>Sample Preparation</strong></li>
<li>Use dedicated workspaces for pre- and post-PCR steps</li>
<li>Implement strict cleaning protocols between samples</li>
<li>
<p>Use UMI-based library preparation methods</p>
</li>
<li>
<p><strong>Sequencing Considerations</strong></p>
</li>
<li>Use unique dual indexing to minimize index hopping</li>
<li>Implement proper library pooling strategies</li>
<li>
<p>Monitor sequencing metrics for contamination indicators</p>
</li>
<li>
<p><strong>Data Analysis</strong></p>
</li>
<li>Always calculate and report contamination estimates</li>
<li>Use UMI-based variant calling for better specificity</li>
<li>Implement duplicate read removal and proper alignment</li>
</ol>
<h3 id="computational-mitigation">Computational Mitigation</h3>
<ol>
<li><strong>Background Subtraction</strong></li>
<li>Subtract estimated contamination from observed frequencies</li>
<li>
<p>Use control samples to estimate background rates</p>
</li>
<li>
<p><strong>Statistical Filtering</strong></p>
</li>
<li>Apply stricter significance thresholds for low-frequency variants</li>
<li>
<p>Use contamination-aware statistical models</p>
</li>
<li>
<p><strong>Quality-Based Filtering</strong></p>
</li>
<li>Filter variants based on UMI family support</li>
<li>Remove variants with suspicious read distribution patterns</li>
</ol>
<h2 id="conclusion">Conclusion</h2>
<p>Contamination analysis is essential for reliable ctDNA/MRD detection. Key takeaways:</p>
<ol>
<li><strong>Monitor contamination continuously</strong>: Even small amounts can affect MRD detection</li>
<li><strong>Use UMI-based methods</strong>: They provide better contamination resistance</li>
<li><strong>Set appropriate QC thresholds</strong>: Based on your clinical requirements</li>
<li><strong>Validate your assay</strong>: Regular contamination testing ensures reliable results</li>
</ol>
<h3 id="next-steps">Next Steps</h3>
<ol>
<li><strong>Run contamination analysis</strong>: Use <code>precise-mrd eval-contamination</code> on your data</li>
<li><strong>Implement QC workflows</strong>: Set up automated contamination monitoring</li>
<li><strong>Optimize your protocol</strong>: Use these insights to improve your lab processes</li>
<li><strong>Validate with controls</strong>: Always include positive and negative controls</li>
</ol>
<hr />
<p>This tutorial covered the essential aspects of contamination analysis in ctDNA/MRD. For more advanced topics, explore the <a href="../../">Precise MRD documentation</a> or run the contamination evaluation commands on your own datasets.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        
<div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/altalanta/precise-mrd-mini" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.sections", "navigation.expand", "navigation.path", "navigation.top", "search.highlight", "search.share", "content.code.copy", "content.code.annotate"], "search": "../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>